clang/lib/Parse/ParseStmt.cpp

```diff
@@ -2427,3 +2447,47 @@ bool Parser::ParseOpenCLUnrollHintAttribute(ParsedAttributes &Attrs) {
   }
   return true;
 }
+
+bool Parser::ParseWhereClauseFact(WhereClause *WClause) {
+  SourceLocation Loc = Tok.getLocation();
+
+  if (Tok.is(tok::identifier) && NextToken().is(tok::colon)) {
+    IdentifierInfo *VarName = Tok.getIdentifierInfo();
+
+    // Consume the identifier.
+    ConsumeToken();
+    // Consume the ':' token.
+    ConsumeToken();
+
+    // Parse bounds expression.
+    ExprResult BoundsRes(ParseBoundsExpression());
+    return Actions.ActOnBoundsFact(WClause, VarName, BoundsRes,
+                                   getCurScope(), Loc);
+  }
+
+  ExprResult ExprRes(ParseExpression());
+  return Actions.ActOnRelopFact(WClause, ExprRes, Loc);
+}
+
+WhereClause *Parser::ParseWhereClause() {
+  SourceLocation Loc = Tok.getLocation();
+
+  // Consume the "_Where" token.
+  if (ExpectAndConsume(tok::kw__Where))
+    return nullptr;
+
+  WhereClause *WClause = Actions.ActOnWhereClause(Loc);
```

Comment:
The null check on WClause is deep inside ActOnBoundsFact and ActOnRelopFact. Maybe it could move here?